roachtest,sqlinstance: fix multitenant_upgrade test

rafiss opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/81517
fixes https://github.com/cockroachdb/cockroach/issues/81506

### roachtest: fix ordering of multitenant upgrade test step

The tenant must be initialized by a system tenant using an old binary in
order to be running on an old binary itself.

Also, we no longer finalize the secondary tenant upgrade before finalizing the
system tenant upgrade.

### sqlinstance: use the correct timestamp when populating the cache

Previously, the sql instance cache would be populated using the
timestamp of the initial rangefeed scan. This was a bug since the
timestamp is used to determine which sql instance ID to use when there
are two IDs for the same address (i.e., it should use the most recent
one).

This caused problems if a tenant node was shutdown, then started with
the same address. Other layers like DistSQL would try to route requests
to an old sql instance, which would actually be the same as the
requesting node.

Release note: None

Release justification: test only change, and critical bug fix

rafiss committed:
roachtest: fix multitenant_upgrade test upgrade order

The tenant must be initialized by a system tenant using an old binary in
order to be running on an old binary itself.

Also, we no longer finalize the secondary tenant upgrade before finalizing the
system tenant upgrade.

Release note: None

Release justification: test only change

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87111)
<!-- Reviewable:end -->

rafiss committed:
sqlinstance: use the correct timestamp when populating the cache

Previously, the sql instance cache would be populated using the
timestamp of the initial rangefeed scan. This was a bug since the
timestamp is used to determine which sql instance ID to use when there
are two IDs for the same address (i.e., it should use the most recent
one).

This caused problems if a tenant node was shutdown, then started with
the same address. Other layers like DistSQL would try to route requests
to an old sql instance, which would actually be the same as the
requesting node.

No release note since this bug didn't affect anything prior to v22.2.

Release justification: high priority bug fix.

Release note: None

ajwerner commented on pkg/sql/sqlinstance/instancestorage/instancereader.go:
```diff
@@ -162,6 +162,7 @@ func (r *Reader) maybeStartRangeFeed(ctx context.Context) *rangefeed.RangeFeed {
 		updateCacheFn,
 		rangefeed.WithInitialScan(initialScanDoneFn),
 		rangefeed.WithOnInitialScanError(initialScanErrFn),
+		rangefeed.WithRowTimestampInInitialScan(true),
```
honestly, this is the problem with functional options :(

ajwerner reviewed (approved):
LGTM!

rafiss commented:
tftr!

bors r=ajwerner
